pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -345,14 +345,19 @@ impl InMemoryLayer {
 
         let cursor = inner.file.block_cursor();
 
-        let mut keys: Vec<(&Key, &VecMap<Lsn, u64>)> = inner.index.iter().collect();
+        // Sort the keys because delta layer writer expects them sorted.
+        //
+        // NOTE: this sort can take up significant time if the layer has millions of
+        //       keys. To speed up all the comparisons we convert the key to i128 and
+        //       keep the value as a reference.
+        let mut keys: Vec<_> = inner.index.iter().map(|(k, m)| (k.to_i128(), m)).collect();
         keys.sort_by_key(|k| k.0);
```

Comment:
Did you consider https://doc.rust-lang.org/stable/std/primitive.slice.html#method.sort_unstable_by_key? Since we sort by what used to be hashmap (IIRC) keys, we should have no duplicates, thus can use the unstable sort.